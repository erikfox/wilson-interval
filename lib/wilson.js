'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

exports.default = function (f, n, c) {
    var N = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
    var continuity = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : false;


    var p = f / n; // proportion of positive outcomes

    if (N) {
        // if population size given:
        var v = Math.sqrt(1 - n / N); // determine scale factor
        n = n / v; // and adjust sample size
    }

    c = c || 0.95; // confidence level, defaults to 95%

    var z = (0, _pnormaldist2.default)(1 - (1 - c) / 2); // calculate z-score: http://www.evanmiller.org/how-not-to-sort-by-average-rating.html

    return continuity ? wilsonIntervalWithContinuity(p, n, z) : wilsonInterval(p, n, z);
};

var _pnormaldist = require('pnormaldist');

var _pnormaldist2 = _interopRequireDefault(_pnormaldist);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Standard Wilson score interval
function wilsonInterval(p, n, z) {

    var high = (p + z * z / (2 * n) + z * Math.sqrt(p * (1 - p) / n + z * z / (4 * (n * n)))) / (1 + z * z / n);
    var low = (p + z * z / (2 * n) - z * Math.sqrt(p * (1 - p) / n + z * z / (4 * (n * n)))) / (1 + z * z / n);
    var center = (p + z * z / (2 * n)) / (1 + z * z / n);

    return {
        high: high,
        center: center,
        low: low
    };
}

// Wilson score interval with continuity correction
function wilsonIntervalWithContinuity(p, n, z) {

    var high = (2 * n * p + z * z + (z * Math.sqrt(z * z - 1 / n + 4 * n * p * (1 - p) - (4 * p - 2)) + 1)) / (2 * (n + z * z));
    var low = (2 * n * p + z * z - (z * Math.sqrt(z * z - 1 / n + 4 * n * p * (1 - p) + (4 * p - 2)) + 1)) / (2 * (n + z * z));
    var center = (2 * n * p + z * z) / (2 * (n + z * z));

    return {
        high: high,
        center: center,
        low: low
    };
}

/*
 * credit: Sean Wallis, Survey of English Usage, University College of London
 * source: https://corplingstats.wordpress.com/2012/04/30/inferential-statistics/
 */
;