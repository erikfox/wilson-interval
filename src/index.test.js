import wilson from '.';

describe('wilson()', () => {
  describe('returns accurate results', () => {
    test('wilson(5, 100)', () => {
      const result = wilson(5, 100);
      expect(result).toMatchSnapshot();
    });

    test('wilson(5, 100, 227)', () => {
      const result = wilson(5, 100, 227);
      expect(result).toMatchSnapshot();
    });
  });

  describe('accepts options', () => {
    test('wilson(5, 100, false, { precision: 6 })', () => {
      const result = wilson(5, 100, false, { precision: 6 });
      expect(result).toMatchSnapshot();
    });

    test('wilson(5, 100, 227, { confidence: 0.9 })', () => {
      const result = wilson(5, 100, 227, { confidence: 0.9 });
      expect(result).toMatchSnapshot();
    });
  });
});
