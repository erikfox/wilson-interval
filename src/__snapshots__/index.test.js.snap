// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wilson() accepts options wilson(5, 100, 227, { confidence: 0.9 }) 1`] = `
Object {
  "center": "0.058925960344442384957",
  "high": "0.090892465800857729375",
  "low": "0.02695945488802704054",
}
`;

exports[`wilson() accepts options wilson(5, 100, false, { precision: 6 }) 1`] = `
Object {
  "center": "0.0666474",
  "high": "0.111751",
  "low": "0.0215437",
}
`;

exports[`wilson() returns accurate results wilson(5, 100) 1`] = `
Object {
  "center": "0.066647073981204927863",
  "high": "0.11175046869375655694",
  "low": "0.021543679268653298792",
}
`;

exports[`wilson() returns accurate results wilson(5, 100, 227) 1`] = `
Object {
  "center": "0.062568829500944323976",
  "high": "0.10110044321850152858",
  "low": "0.024037215783387119365",
}
`;
